dartagnan/src/test/java/com/dat3m/dartagnan/spirv/header/SpirvHeaderSingleTest.java

```diff
@@ -0,0 +1,132 @@
+package com.dat3m.dartagnan.spirv.header;
+
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.parsers.program.ParserSpirv;
+import com.dat3m.dartagnan.program.Program;
+import com.dat3m.dartagnan.program.memory.Location;
+import com.dat3m.dartagnan.program.specification.AssertBasic;
+import com.dat3m.dartagnan.program.specification.AssertCompositeAnd;
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.CharStreams;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class SpirvHeaderSingleTest {
+    public SpirvHeaderSingleTest() {
+    }
+
+    @Test
+    public void testComplexSpecHeader() {
```

Comment:
I would make them a bit more precise and add a few test cases here. Something like this, maybe


@Test
    public void testAssertionScalarLeft() {
        // given
        String input = """
                ; @Output: forall %v1==1 and %v2>2 and %v3<3
                """;

        // when
        Program program = parse(input + spvBody);
        AssertCompositeAnd ast = (AssertCompositeAnd)program.getSpecification();

        // then
        assertEquals(ASSERT_TYPE_FORALL, ast.getType());

        AssertBasic ast1 = (AssertBasic) ((AssertCompositeAnd)ast.getLeft()).getLeft();
        AssertBasic ast2 = (AssertBasic) ((AssertCompositeAnd)ast.getLeft()).getRight();
        AssertBasic ast3 = (AssertBasic) ast.getRight();

        IntegerType int64 = TYPE_FACTORY.getIntegerType(64);

        assertEquals(EXPR_FACTORY.makeValue(1, int64), ast1.getRight());
        assertEquals(EXPR_FACTORY.makeValue(2, int64), ast2.getRight());
        assertEquals(EXPR_FACTORY.makeValue(3, int64), ast3.getRight());

        assertEquals("%v1", ((MemoryObject) ast1.getLeft()).getCVar());
        assertEquals("%v2", ((MemoryObject) ast2.getLeft()).getCVar());
        assertEquals("%v3", ((MemoryObject) ast3.getLeft()).getCVar());
    }

    @Test
    public void testAssertionScalarRight() {
        // given
        String input = """
                ; @Output: forall 1!=%v1 and 2>%v2 and 3<%v3
                """;

        // when
        Program program = parse(input + spvBody);
        AssertCompositeAnd ast = (AssertCompositeAnd)program.getSpecification();

        // then
        assertEquals(ASSERT_TYPE_FORALL, ast.getType());

        AssertBasic ast1 = (AssertBasic) ((AssertCompositeAnd)ast.getLeft()).getLeft();
        AssertBasic ast2 = (AssertBasic) ((AssertCompositeAnd)ast.getLeft()).getRight();
        AssertBasic ast3 = (AssertBasic) ast.getRight();

        IntegerType int64 = TYPE_FACTORY.getIntegerType(64);

        assertEquals(EXPR_FACTORY.makeValue(1, int64), ast1.getLeft());
        assertEquals(EXPR_FACTORY.makeValue(2, int64), ast2.getLeft());
        assertEquals(EXPR_FACTORY.makeValue(3, int64), ast3.getLeft());

        assertEquals("%v1", ((MemoryObject) ast1.getRight()).getCVar());
        assertEquals("%v2", ((MemoryObject) ast2.getRight()).getCVar());
        assertEquals("%v3", ((MemoryObject) ast3.getRight()).getCVar());
    }

    @Test
    public void testAssertionVectorLeft() {
        // given
        String input = """
                ; @Output: forall %v4[0]<=1 and %v4[1]>=2 and %v4[2]==3
                """;

        // when
        Program program = parse(input + spvBody);
        AssertCompositeAnd ast = (AssertCompositeAnd)program.getSpecification();

        // then
        assertEquals(ASSERT_TYPE_FORALL, ast.getType());

        AssertBasic ast1 = (AssertBasic) ((AssertCompositeAnd)ast.getLeft()).getLeft();
        AssertBasic ast2 = (AssertBasic) ((AssertCompositeAnd)ast.getLeft()).getRight();
        AssertBasic ast3 = (AssertBasic) ast.getRight();

        IntegerType int64 = TYPE_FACTORY.getIntegerType(64);

        assertEquals(EXPR_FACTORY.makeValue(1, int64), ast1.getRight());
        assertEquals(EXPR_FACTORY.makeValue(2, int64), ast2.getRight());
        assertEquals(EXPR_FACTORY.makeValue(3, int64), ast3.getRight());

        assertEquals("%v4[0]", ((Location) ast1.getLeft()).getName());
        assertEquals("%v4[1]", ((Location) ast2.getLeft()).getName());
        assertEquals("%v4[2]", ((Location) ast3.getLeft()).getName());

        assertEquals(0, ((Location) ast1.getLeft()).getOffset());
        assertEquals(8, ((Location) ast2.getLeft()).getOffset());
        assertEquals(16, ((Location) ast3.getLeft()).getOffset());
    }

    @Test
    public void testAssertionValueRight() {
        // given
        String input = """
                ; @Output: forall 1<=%v4[0] and 2>=%v4[1] and 3==%v4[2]
                """;

        // when
        Program program = parse(input + spvBody);
        AssertCompositeAnd ast = (AssertCompositeAnd)program.getSpecification();

        // then
        assertEquals(ASSERT_TYPE_FORALL, ast.getType());

        AssertBasic ast1 = (AssertBasic) ((AssertCompositeAnd)ast.getLeft()).getLeft();
        AssertBasic ast2 = (AssertBasic) ((AssertCompositeAnd)ast.getLeft()).getRight();
        AssertBasic ast3 = (AssertBasic) ast.getRight();

        IntegerType int64 = TYPE_FACTORY.getIntegerType(64);

        assertEquals(EXPR_FACTORY.makeValue(1, int64), ast1.getLeft());
        assertEquals(EXPR_FACTORY.makeValue(2, int64), ast2.getLeft());
        assertEquals(EXPR_FACTORY.makeValue(3, int64), ast3.getLeft());

        assertEquals("%v4[0]", ((Location) ast1.getRight()).getName());
        assertEquals("%v4[1]", ((Location) ast2.getRight()).getName());
        assertEquals("%v4[2]", ((Location) ast3.getRight()).getName());

        assertEquals(0, ((Location) ast1.getRight()).getOffset());
        assertEquals(8, ((Location) ast2.getRight()).getOffset());
        assertEquals(16, ((Location) ast3.getRight()).getOffset());
    }

    // TODO: Same for an array with different element with

    // TODO: Same for a structure

Reply:
Added. Also I found as the IntegerType(64) is hardcoded in TypeFactory, I removed the bv32 test.